Source/SwiftLintBuiltInRules/Rules/Lint/PrivateSwiftUIStatePropertyRule.swift

```diff
@@ -0,0 +1,137 @@
+import SwiftSyntax
+
+/// Require that any state properties in SwiftUI be declared as private
+///
+/// State properties should only be accessible from inside the View's body, or from methods called by it
+struct PrivateSwiftUIStatePropertyRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule {
+    var configuration = SeverityConfiguration<Self>(.warning)
+
+    static let description = RuleDescription(
+        identifier: "private_swiftui_state",
+        name: "Private SwiftUI @State Properties",
+        description: "SwiftUI's state properties should be private",
+        kind: .lint,
+        nonTriggeringExamples: [
+            Example("""
+            struct ContentView: View {
+                @State private var isPlaying: Bool = false
+            }
+            """),
+            Example("""
+            struct ContentView: View {
+                @State fileprivate var isPlaying: Bool = false
+            }
+            """),
+            Example("""
+            struct ContentView: View {
+                var isPlaying = false
+            }
+            """),
+            Example("""
+            struct ContentView: View {
+                @StateObject var foo = Foo()
+            }
+            """),
+            Example("""
+            struct Foo {
+                @State var bar = false
+            }
+            """),
+            Example("""
+            class Foo: ObservableObject {
+                @State var bar = Bar()
+            }
+            """),
+            Example("""
+            extension MyObject {
+                struct ContentView: View {
+                    @State private var isPlaying: Bool = false
+                }
+            }
+            """)
+        ],
+        triggeringExamples: [
+            Example("""
+            struct ContentView: View {
+                @State ↓var isPlaying: Bool = false
+            }
+            """),
+            Example("""
+            final class ContentView: View {
+                @State ↓var isPlaying: Bool = false
+            }
+            """),
+            Example("""
+            extension MyObject {
+                struct ContentView: View {
+                    @State ↓var isPlaying: Bool = false
+                }
+            }
+            """)
+        ]
+    )
+
+    func makeVisitor(file: SwiftLintFile) -> ViolationsSyntaxVisitor {
+        Visitor(viewMode: .sourceAccurate)
+    }
+}
+
+private extension PrivateSwiftUIStatePropertyRule {
+    final class Visitor: ViolationsSyntaxVisitor {
+        override var skippableDeclarations: [DeclSyntaxProtocol.Type] {
+            [EnumDeclSyntax.self, FunctionDeclSyntax.self, ProtocolDeclSyntax.self, VariableDeclSyntax.self]
```

Comment:
Even though unlikely, types can be defined almost everywhere, especially also in enum, function and variable declarations. That said, protocols are actually the only declaration type that can be ignored.